src/irc_bridge.py

```diff
@@ -0,0 +1,167 @@
+        """
+        if not charm_state.synapse_config.enable_irc_bridge:
+            return
+        if charm_state.irc_bridge_datasource is None:
+            ops.MaintenanceStatus("Waiting for irc bridge db relation")
+            return
+        container = self._charm.unit.get_container(synapse.SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            self._charm.unit.status = ops.MaintenanceStatus("Waiting for Synapse pebble")
+            return
+        irc_service = container.get_services(IRC_SERVICE_NAME)
+        if irc_service:
+            logger.debug("%s service already exists, skipping", IRC_SERVICE_NAME)
+            return
+        self._enable_irc_bridge(charm_state)
+        event.add_status(ops.ActiveStatus())
+
+    def _enable_irc_bridge(self, charm_state: CharmState) -> None:
+        """Enable irc service.
+
+        The required steps to enable the IRC bridge are:
+         - Create the IRC bridge configuration file.
+         - Create the IRC bridge registration file.
+         - Generate a PEM file for the IRC bridge.
+         - Finally, add IRC bridge pebble layer.
+
+        Args:
+            charm_state: Instance of CharmState.
+
+        """
+        container = self._charm.unit.get_container(synapse.SYNAPSE_CONTAINER_NAME)
+        if not container.can_connect():
+            self._charm.unit.status = ops.MaintenanceStatus("Waiting for Synapse pebble")
+            return
+        self._charm.model.unit.status = ops.MaintenanceStatus("Configuring IRC bridge")
+        server_name = charm_state.synapse_config.server_name
+        db_connect_string = self._get_db_connection(charm_state)
+        synapse.create_irc_bridge_config(
+            container=container, server_name=server_name, db_connect_string=db_connect_string
+        )
+        synapse.create_irc_bridge_app_registration(container=container)
+        self._create_pem_file(container=container)
+        pebble.replan_irc_bridge(container)
+        pebble.restart_synapse(charm_state, container)
+        self._charm.model.unit.status = ops.ActiveStatus()
+
+    def _create_pem_file(self, container: ops.model.Container) -> None:
+        """Create a PEM file for the IRC bridge.
+
+        Args:
+            container: The container to create the PEM file in.
+
+        Raises:
+            PEMCreateError: If the PEM file creation fails.
+        """
+        self._charm.model.unit.status = ops.MaintenanceStatus("Creating PEM file for IRC bridge")
+        pem_create_command = [
+            "/bin/bash",
+            "-c",
+            "[[ -f /data/config/irc_passkey.pem ]] || "
+            + "openssl genpkey -out /data/config/irc_passkey.pem "
+            + "-outform PEM -algorithm RSA -pkeyopt rsa_keygen_bits:2048",
+        ]
+        logger.info("Creating PEM file for IRC bridge.")
+        try:
+            exec_process = container.exec(
+                pem_create_command,
+                environment={},
```

Comment:
I'm wondering if `environment={}` intends to do something?

Reply:
I don't recall placing it, let's say it is there to see if people review PRs thoroughly - you get the bonus points :tada: and thanks for spotting.